roachtest: acceptance/multitenant failed

roachtest.acceptance/multitenant [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2988247&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2988247&tab=artifacts#/acceptance/multitenant) on master @ [adb98c20a536c80885b3eddb7119248a6549d29f](https://github.com/cockroachdb/cockroach/commits/adb98c20a536c80885b3eddb7119248a6549d29f):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: artifacts/acceptance/multitenant/run_1
	sql_runner.go:59,multitenant_upgrade.go:431,multitenant.go:47,acceptance.go:98,test_runner.go:784: error executing 'CREATE TABLE foo (id INT PRIMARY KEY, v STRING)': dial tcp 127.0.0.1:36258: connect: connection refused
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
# From https://go.crdb.dev/p/roachstress, perhaps edited lightly.
caffeinate ./roachstress.sh acceptance/multitenant
```
</p>
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*acceptance/multitenant.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/cli/mt_start_sql.go
+++ b/pkg/cli/mt_start_sql.go
@@ -137,5 +137,6 @@ func runStartSQL(cmd *cobra.Command, args []string) error {
 	ch := make(chan os.Signal, 1)
 	signal.Notify(ch, drainSignals...)
 	sig := <-ch
+	log.Flush()
 	return errors.Newf("received signal %v", sig)
 }

--- a/pkg/cmd/roachtest/multitenant.go
+++ b/pkg/cmd/roachtest/multitenant.go
@@ -53,8 +53,7 @@ func runAcceptanceMultitenant(ctx context.Context, t *test, c *cluster) {
 	t.Status("stopping the server ahead of checking for the tenant server")
 
 	// Stop the server, which also ensures that log files get flushed.
-	cancel()
-	<-tenant.errCh
+	tenant.stop(ctx, t, c)
 
 	t.Status("checking log file contents")
 

--- a/pkg/cmd/roachtest/multitenant_upgrade.go
+++ b/pkg/cmd/roachtest/multitenant_upgrade.go
@@ -104,7 +104,7 @@ func (tn *tenantNode) start(ctx context.Context, t *test, c *cluster, binary str
 	u.Host = internalUrls[0] + ":" + strconv.Itoa(tn.sqlPort)
 	tn.pgURL = u.String()
 
-	// The tenant is usually responsible ~right away, but it
+	// The tenant is usually responsive ~right away, but it
 	// has on occasions taken more than 3s for it to connect
 	// to the KV layer, and it won't open the SQL port until
 	// it has.

--- a/pkg/cmd/roachtest/multitenant_upgrade.go
+++ b/pkg/cmd/roachtest/multitenant_upgrade.go
@@ -20,6 +20,7 @@ import (
 	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
 	"github.com/cockroachdb/cockroach/pkg/util/version"
 	"github.com/stretchr/testify/require"
 )
@@ -102,14 +103,32 @@ func (tn *tenantNode) start(ctx context.Context, t *test, c *cluster, binary str
 	require.NoError(t, err)
 	u.Host = internalUrls[0] + ":" + strconv.Itoa(tn.sqlPort)
 	tn.pgURL = u.String()
-	c.l.Printf("sql server for tenant %d should be running at %s", tn.tenantID, tn.pgURL)
 
-	const warmupPeriod = 3 * time.Second
-	select {
-	case err := <-tn.errCh:
+	// The tenant is usually responsible ~right away, but it
+	// has on occasions taken more than 3s for it to connect
+	// to the KV layer, and it won't open the SQL port until
+	// it has.
+	if err := retry.ForDuration(45*time.Second, func() error {
+		select {
+		case <-ctx.Done():
+			t.Fatal(ctx.Err())
+		case <-tn.errCh:
+			t.Fatal(err)
+		default:
+		}
+
+		db, err := gosql.Open("postgres", tn.pgURL)
+		if err != nil {
+			return err
+		}
+		defer db.Close()
+		_, err = db.ExecContext(ctx, `SELECT 1`)
+		return err
+	}); err != nil {
 		t.Fatal(err)
-	case <-time.After(warmupPeriod):
 	}
+
+	c.l.Printf("sql server for tenant %d running at %s", tn.tenantID, tn.pgURL)
 }
 
 // runMultiTenantUpgrade exercises upgrading tenants and their host cluster.

